README.md
# AVR Net enc28j60

Library written in C for AVR microcontroller. This code is based on other public repository which i do not remember where it is.
If someone will recognize code which is not mine and will know repository i will glad add reference here.

## Versions

Current version **0.1.4**

```console
git clone -b 0.1.4 git@github.com:Triplkrypl/avr-net-enc28j60.git
```

Development version **0.2**

[Changes](CHANGES.md)

## Requires

* Hardware
    * Some AVR microcontroller of course
    * enc28j60 chip
* Software
    * [AVR library](https://www.nongnu.org/avr-libc/)

## Usage

### Examples

[Tcp synchronous not persistent client](example/tcp_synch_client/main.c)

[Tcp asynchronous not persistent client](example/tcp_asynch_client/main.c)

### General usage

```c
// define SPI pins connected to enc28j60 SI, SO, SCK pins
// SPI pins can changes on different AVR processors look into chip datasheet where SPI pins are
// this example is for AVR atmega328p
#define ENC28J60_DDR DDRB
#define ENC28J60_PORT PORTB
#define ENC28J60_SI_PIN_DDR DDB3
#define ENC28J60_SI_PIN PORTB3
#define ENC28J60_SO_PIN_DDR DDB4
#define ENC28J60_SCK_PIN_DDR DDB5
#define ENC28J60_SCK_PIN PORTB5

// define rest digitals I/O pins connected to enc28j60 RESET, INT, CS pins
// this pins have to be in same DDR and PORT as SPI pins
#define ENC28J60_RESET_PIN_DDR DDB0
#define ENC28J60_RESET_PIN PORTB0
#define ENC28J60_INT_PIN_DDR DDB1
#define ENC28J60_INT_PIN PORTB1
#define ENC28J60_CS_PIN_DDR DDB2
#define ENC28J60_CS_PIN PORTB2

// define IP, MAC address for you AVR
#define NET_IP 192, 168, 0, 150
#define NET_MAC 0x15, 0x8, 0x45, 0x89, 0x69, 0x99
// define RAM buffer size for network packets
// this define maximum size of incoming packet,
// define maximum size of sended packet,
// and define tcp max segment size
// bigger value allow send and receive bigger packet (UDP datagram) and better performance for TCP
// lower value save RAM, you can create bigger buffer than AVR RAM or to big than no space for application :-)
// minimum recommended value is 600 bytes, in IPV4 standard minimum IP datagram size is 576 bytes
// you can use smaller buffer, if you sure than you do not receive bigger packets
#define NET_BUFFER_SIZE 600
// define size of arp cache array, default value 5
// arp requests for ip are cached in RAM
// defined zero size turn off cache and code for cache is not included in compilation
// bigger value allow save more ip in cache is good if your AVR communicate with more hosts
// lower value is good for more dynamic network in smaller cache will often rewrite unused ip in cache
#define NET_ARP_CACHE_SIZE 5
// define for tcp.c include
// define how many tcp connection can live in AVR (server/client connections)
// if number connections is equal as limit all new incoming TCP connections is dropped
// bigger value allow more connected hosts at one time
// lower value is RAM free
#define TCP_MAX_CONNECTIONS 2
// define for http.c include
// define which tcp port is used for http listener
#define HTTP_SERVER_PORT 80
// definition of maximum request parts length, whole received request can be bigger than
// NET_BUFFER_SIZE because request can be received in multiple tcp packet but be ware out of memory
// define maximum request url length in bytes without host, example "/some/url?somequery"
#define HTTP_MAX_URL_LENGTH 50
// define maximum request headers length without first request header line where is method...
#define HTTP_MAX_HEADER_ROWS_LENGTH 50
// define maximum request body length
#define HTTP_MAX_DATA_LENGTH 50
// turn on/off callback from tcp.c, if http.c is included
// http.c define all callbacks from tcp.c so can not be used without define HTTP_TCP_INCLUDED as 1
// if tcp callbacks is turn on have to be defined in application as if tcp.c is included
// if you want use only http protocol you do not need this define and use default value 0
// if you want implement other protocol which used tcp and use http to, turn tcp callback on
// all tcp callback is not call on packet related to http protocol, this packets is handled inside http.c
#define HTTP_TCP_INCLUDED 0
// you can define end of all outgoing http header rows by default is used windows row ending,
// change only if you know than application on other side accept other row ending
//#define HTTP_HEADER_ROW_BREAK "\r\n"

// you have to define your cpu frequency because of delay.h library
#define F_CPU 16000000UL

// include icmp.c if you want your AVR to response PING
#include "src/icmp.c"
// include udp.c if you want handle incoming UDP datagram or send them
//#include "src/udp.c"
// include tcp.c if you want handle incoming TCP connection or connect somewhere
//#include "src/tcp.c"
// include http.c if you want handle incoming HTTP requests
//#include "src/http.c"
// include one of icmp.c or udp.c or tcp.c or http.c,
// otherwise AVR will only response on ARP and it is little useless :-)
// not include tcp.c and http.c both if you want use both correctly,
// include http.c and before define HTTP_TCP_INCLUDED as 1
// you have to include network.c (main include for library)
#include "src/network.c"

int main(){
 // at start of main function call NetInit for init enc28j60 chip init memory for cache...
 NetInit();
 for(;;){
  // in application loop everytime call NetHandleNetwork for handle incoming packet from ethernet
  NetHandleNetwork();
 }
}
```

### Callback functions

If you include **tcp.c**, you have to **define** functions callback **TcpOnNewConnection**, **TcpOnConnect**, **TcpOnIncomingData**, **TcpOnDisconnect**
You also have to define this callbacks if you include http.c and define HTTP_TCP_INCLUDED as 1

```c
// function called if new tcp client is connecting into AVR, it is like firewall,
// library listen on all TCP ports you can specify witch port is accepted or not
// you have to return:
// NET_HANDLE_RESULT_DROP drop new connection without any response for sender
// NET_HANDLE_RESULT_OK accept new connection and allow established connection
// NET_HANDLE_RESULT_REJECT not allow established connection and library send icmp unreachable packet
unsigned char TcpOnNewConnection(const unsigned char connectionId){
 if(TcpGetConnection(connectionId)->port != 80){
  return NET_HANDLE_RESULT_DROP;
 }
 return NET_HANDLE_RESULT_OK;
}

// function called after successful established any connection
// server connection and client connection created by TcpConnect function
// purpose this callback s init any data or memory for new connection before send data
void TcpOnConnect(const unsigned char connectionId){
 // if you listen on more than one ports you always check port for execute correct logic
 // also if you use client connection from AVR anywhere,
 // check here you client connection id for handle asynchronous communication
 if(TcpGetConnection(connectionId)->port == 80){
 }
}

// function called always if AVR receive any data from any connection,
// server receive data from client or client receive asynchronous data
// data catched by synchronous wait is not send in this function
void TcpOnIncomingData(const unsigned char connectionId, const unsigned char *data, unsigned short dataLength){
 // here always check port if you listen on multiple ports or use client connection with asynchronous responses
 // so you can deliver incoming data in to correct logic
 if(TcpGetConnection(connectionId)->port == 80){
 }
}

// function called after closed any connection passively from other side or actively by TcpDisconnect function
// it is necessary for correct clear application memory after any connection ends
// if you prepare any memory in TcpOnConnect callback clear it here
void TcpOnDisconnect(const unsigned char connectionId){
 // always check port or client connection id here,
 // if you for different connection type prepare different memory you have to know what you have to clear
 // use this callback even if you use synchronous wait for response,
 // other side can close connection asynchronously sooner than you expect
 if(TcpGetConnection(connectionId)->port == 80){
 }
}
```

If you include **udp.c**, you have to **define** function callback **UdpOnIncomingDatagram**

```c
// function called any time if any UPD datagram income into AVR except UDP datagram cathed by synchronous wait
// you have to return:
// NET_HANDLE_RESULT_DROP inform library that your application code nothing do with datagram
// NET_HANDLE_RESULT_OK inform library that your application code do somethink with datagram
// NET_HANDLE_RESULT_REJECT inform library that your application code nothing do with datagram
// after that library send unreachable icmp packet
unsigned char UdpOnIncomingDatagram(const UdpDatagram datagram, const unsigned char *data, unsigned short dataLength){
 if(datagram.port == 5000){
  // do something
  return NET_HANDLE_RESULT_OK;
 }
 return NET_HANDLE_RESULT_DROP;
}
```

If you include **http.c**, you have to **define** function callback **HttpOnIncomingRequest**
```c
// function called any time if http request if fully parsed and ready to be processed
// only request incoming with destination HTTP_SERVER_PORT is allow for parsing and passed in this callback
// response must be send by function HttpSendResponse called inside this callback
// response is not returned as return value because otherwise is dynamic allocation needed
// if HttpSendResponse is not called, library close tcp connection without send any response after callback ends
// any 404 and 500 or other applications errors have to be handled in application
void HttpOnIncomingRequest(const HttpRequest* request){
 if(strcmp(request->method, "GET") == 0 && CharsCmp(request->url, request->urlLength, "/some-url", strlen("/some-url"))){
  // do some logic
  HttpStatus status = {200, "OK"};
  HttpSendResponse(&status, "SomeHeader: :-)\n", strlen("SomeHeader: :-)\n"), "SomeData", strlen("SomeData"));
  return;
 }
 HttpStatus status = {404, "Not Found"};
 HttpSendResponse(&status, 0, 0, 0, 0);
}
```

### Functions

Functions from **tcp.c TcpGetConnection**, **TcpConnect**, **TcpSendData**, **TcpReceiveData**, **TcpDisconnect**

```c
// function return connection structure pointer with information about TCP connection remote ip, mac, port...
// if connection not found by connectionId return 0
const TcpConnection *TcpGetConnection(const unsigned char connectionId);

// function connect as client into server with milliseconds timeout
// return: TCP_INVALID_CONNECTION_ID in any error or after timeout on success return established connectionId
unsigned char TcpConnect(const unsigned char ip[IP_V4_ADDRESS_SIZE], const unsigned short remotePort, const unsigned short timeout);

// function send data into any established connection with timeout for TCP ack packet
// return: 1 on success, 0 on any error
// you can send bigger data than NET_BUFFER_SIZE if you do this data is send in multiple packet
unsigned char TcpSendData(const unsigned char connectionId, const unsigned short timeout, const unsigned char *data, unsigned short dataLength);

// synchronous waiting for data with milliseconds timeout, zero timeout value means infinite waiting,
// in infinite timeout, function ends only if receive data or connection is closed
// after receive data are accessible by double pointer **data parameter
// this function is good if you send data as client and waiting for response from server,
// not use this function in TcpOnIncomingData on parameter connectionId another data will came in next callback call
// receive data by this function will never be send into TcpOnIncomingData callback
// return: 1 on success, 0 on any error
unsigned char TcpReceiveData(const unsigned char connectionId, const unsigned short timeout, unsigned char **data, unsigned short *dataLength);

// actively close any TCP connection, you can disconnect inactive clients as server,
// or client have to notify server that ending communication
// return: 1 on success, 0 on any error
unsigned char TcpDisconnect(const unsigned char connectionId, unsigned short timeout);
```

Functions from **udp.c UdpSendDataMac**, **UdpSendData**, **UdpSendDataTmpPort**, **UdpReceiveData**

```c
// send data into any host, you need know mac address,
// this function is good in UdpOnIncomingDatagram if you sending response as UDP server
// return:
// on success port used as srcPort in sended UDP datagram (port parameter)
// on fail return 0
unsigned short UdpSendDataMac(const unsigned char *mac, const unsigned char *ip, const unsigned short remotePort, const unsigned short port, const unsigned char *data, const unsigned short dataLength);

// send data into any host, you not need know mac address, mac address is lookup by ARP
// return:
// on success port used as srcPort in sended UDP datagram (port parameter)
// on fail return 0
unsigned short UdpSendData(const unsigned char *ip, const unsigned short remotePort, const unsigned short port, const unsigned char *data, const unsigned short dataLength);

// send data into any host, mac address is lookup and you specify only remotePort
// port (srcPort in sended UDP datagram is incrementing UDP port)
// this function is good for sending UDP request as client with synchronous waiting for response,
// you do not need store port value for long time
// return:
// on success port used as srcPort in sended UDP datagram
// (incrementing port, every sended UDP datagram rotate new port,
// also new client Tcp connection will cause incrementing)
// on fail return 0
// if you use this function instantly use UdpReceiveData for catch response with returned port value,
// or you lose it forever
unsigned short UdpSendDataTmpPort(const unsigned char *ip, const unsigned short remotePort, const unsigned char *data, const unsigned short dataLength);

// synchronous wait for data from specify host with timeout
// use it if you pray for response from server or any other host
// catched data by this function is not send into UdpOnIncomingDatagram callback
unsigned char UdpReceiveData(const unsigned char *ip, const unsigned short remotePort, const unsigned port, unsigned short timeout, unsigned char **data, unsigned short *dataLength);
```

Functions from **http.c HttpParseHeaderValue**, **HttpSendResponse**

```c
// function parse value from http header rows, first parameter is http message as source of headers
// second parameter is header row key name for look up
// if header key name are more than once in message first one value is returned
// if header key name is not found in message, structure with zero length and zero pointer value is returned
// example: HttpParseHeaderValue(message, "Host"); with http headers:
// Host: somewhere.org
// SomeHeader: aaaaaaa
//
// will return structure with 13 length and "somewhere.org" value pointing into message structure without ending zero char
const HttpHeaderValue HttpParseHeaderValue(const HttpMessage *message, const unsigned char *header);

// function send http response into network while is some request processed in callback HttpOnIncomingRequest
// function return 1 on success 0 on any error
// can be called only once per callback call and only if request is processed, in other http listener internal state return 0 as error
// if HttpStatus.message is 0 pointer default status message is filed by library
// headers parameter is all additional response header rows example:
// "ContentType: json" HTTP_HEADER_ROW_BREAK
// "Language: en" HTTP_HEADER_ROW_BREAK
// all header rows have to end with HTTP_HEADER_ROW_BREAK string value
// usage of defined constant HTTP_HEADER_ROW_BREAK is important due to consistent rows end sent by library and rows end in application
// headersLength parameter is length of headers parameter in bytes, if zero length is set no additional headers are send and headers parameter is ignored
// data parameter is data send in response body if dataLength is zero no body is send and data parameter is ignored
unsigned char HttpSendResponse(const HttpStatus *status, unsigned char *headers, unsigned short headersLength, unsigned char *data, unsigned short dataLength);
```



CHANGES.md
## 0.1 - 0.2

* Changed **HttpRequest structure** used in **HttpOnIncomingRequest parameter**, some structure **members moved into HttpMessage** which is accessible by pointer in HttpRequest
* Declaration of **HttpParseHeaderValue** has **HttpMessage** as first parameter **instead of HttpRequest**
* Changed behavior **if HttpSendResponse is not called in HttpOnIncomingRequest**, library will **close connection instead** of **send Http 204 response**
* Definition **HTTP_HEADER_ROW_BREAK** added with default value "\r\n", all outgoing http header rows end with this defined value, even application header rows must
* Incoming http header rows stored in **HttpMessage.headers** are with **raw row ending**, "\r\n" sequence not converted into "\n"
* Function HttpSendResponse now not lookup default status message for status code if HttpStatus.message is zero pointer, only default fallback message is used

* Function **TcpReceiveData** parameter **timeout** has different behavior with zero value, it is means infinite timeout for data

## Changes what i made before versioning.

* Fixes:
    * icmp ping with different length works, before library not response with same icmp length as was request but with constant length

* Improvements:
    * tcp can handle multiple established connections
    * synchronous wait for packet if receive different packet than expected will send received packed in to main loop function for handle except drop
    * arp broadcast responses are cached


src/http.c
#ifndef HTTP
#define HTTP

#include <stdio.h>
#include <string.h>

#define TCP_ON_NEW_CONNETION_CALLBACK HttpTcpOnNewConnection
#define TCP_ON_CONNECT_CALLBACK HttpTcpOnConnect
#define TCP_ON_INCOMING_DATA_CALLBACK HttpTcpOnIncomingData
#define TCP_ON_DISCONNECT_CALLBACK HttpTcpOnDisconnect

#include "tcp.c"

#ifndef HTTP_TCP_INCLUDED
#define HTTP_TCP_INCLUDED 0
#endif

#define HTTP_MAX_METHOD_LENGTH 10
#define HTTP_MAX_VERSION_LENGTH 10

#define HTTP_REQUEST_STATE_NO_REQUEST 0
#define HTTP_REQUEST_STATE_START_REQUEST 1
#define HTTP_REQUEST_STATE_METHOD 2
#define HTTP_REQUEST_STATE_URL 3
#define HTTP_REQUEST_STATE_VERSION 4

#define HTTP_STATE_LINUX_END_HEADER 5
#define HTTP_STATE_MAC_END_HEADER 6
#define HTTP_STATE_WIN_END_HEADER 7
#define HTTP_STATE_WIN_END_HEADER2 8
#define HTTP_STATE_HEADER 9
#define HTTP_STATE_END_HEADER 10
#define HTTP_STATE_END_MESSAGE 11

#define HTTP_REQUEST_STATE_REQUEST_HANDLING 12

#ifndef HTTP_HEADER_ROW_BREAK
#define HTTP_HEADER_ROW_BREAK "\r\n"
#endif

typedef struct{
 unsigned short headersLenght;
 unsigned char headers[HTTP_MAX_HEADER_ROWS_LENGTH+1];
 unsigned char version[HTTP_MAX_VERSION_LENGTH+1];
 unsigned char data[HTTP_MAX_DATA_LENGTH];
 unsigned char dataLength;
} HttpMessage;

typedef struct{
 const TcpConnection *connection;
 HttpMessage *message;
 unsigned char method[HTTP_MAX_METHOD_LENGTH+1];
 unsigned char url[HTTP_MAX_URL_LENGTH];
 unsigned char urlLength;
} HttpRequest;

typedef struct{
 unsigned short code;
 const char *message;
} HttpStatus;

typedef struct{
 unsigned short length;
 unsigned char *value;
} HttpHeaderValue;

#if HTTP_TCP_INCLUDED == 1
unsigned char TcpOnNewConnection(const unsigned char connectionId);
void TcpOnConnect(const unsigned char connectionId);
void TcpOnIncomingData(const unsigned char connectionId, const unsigned char *data, unsigned short dataLength);
void TcpOnDisconnect(const unsigned char connectionId);
#endif
const HttpHeaderValue HttpParseHeaderValue(const HttpMessage *message, const unsigned char *header);
void HttpOnIncomingRequest(const HttpRequest *request);

static unsigned short incomingPosition;
static unsigned char incomingRequestConnectionId;
static unsigned char incomingRequestState;
static HttpMessage incomingMessage;
static HttpRequest incomingRequest;

//*****************************************************************************************
//
// Function : HttpHeadersPutChar
// Description : put chat into http message structure why parsing headers string
//
//*****************************************************************************************
static unsigned char HttpHeadersPutChar(unsigned char ch){
 if((ch == '\n' || ch  == '\r') && incomingMessage.headersLenght == 0){
  return 1;
 }
 if(incomingMessage.headersLenght >= HTTP_MAX_HEADER_ROWS_LENGTH){
  return 0;
 }
 incomingMessage.headers[incomingMessage.headersLenght] = ch;
 incomingMessage.headersLenght++;
 return 1;
}

//*****************************************************************************************
//
// Function : HttpMessagePutData
// Description : function copy chars from data pointer into global HttpMessage object
//
//*****************************************************************************************
static void HttpMessagePutData(const unsigned char *data, unsigned short dataLength, const unsigned short dataPosition){
 if(dataLength == 0){
  return;
 }
 dataLength -= dataPosition;
 data += dataPosition;
 memcpy(incomingMessage.data + incomingMessage.dataLength, data, dataLength);
 incomingMessage.dataLength += dataLength;
}

//*****************************************************************************************
//
// Function : HttpParseHeader
// Description : function take char from incoming data and detect end of http header
// Return 1 on success 0 on header to long
//
//*****************************************************************************************
static unsigned char HttpParseHeader(const unsigned char ch, unsigned char *headerState){
 // parse headers end
 if(*headerState == HTTP_STATE_MAC_END_HEADER){
  if(ch == '\n'){
   *headerState = HTTP_STATE_WIN_END_HEADER;
   return HttpHeadersPutChar(ch);
  }
  if(ch == '\r'){
   *headerState = HTTP_STATE_END_HEADER;
   incomingMessage.dataLength = 0;
   incomingMessage.headers[incomingMessage.headersLenght] = 0;
   return 1;
  }
  *headerState = HTTP_STATE_HEADER;
 }
 if(*headerState == HTTP_STATE_LINUX_END_HEADER || *headerState == HTTP_STATE_WIN_END_HEADER2){
  if(ch == '\n'){
   *headerState = HTTP_STATE_END_HEADER;
   incomingMessage.dataLength = 0;
   incomingMessage.headers[incomingMessage.headersLenght] = 0;
   return 1;
  }
  *headerState = HTTP_STATE_HEADER;
 }
 if(*headerState == HTTP_STATE_WIN_END_HEADER){
  if(ch == '\r'){
   *headerState = HTTP_STATE_WIN_END_HEADER2;
   return 1;
  }
  *headerState = HTTP_STATE_HEADER;
 }
 // parse header rows
 if(*headerState == HTTP_STATE_HEADER){
  if(ch == '\r'){
   *headerState = HTTP_STATE_MAC_END_HEADER;
  }
  if(ch == '\n'){
   *headerState = HTTP_STATE_LINUX_END_HEADER;
  }
  return HttpHeadersPutChar(ch);
 }
 return 0;
}

//*****************************************************************************************
//
// Function : HttpSendResponseHeader
// Description : build http response header and send it into tcp connection
//
//*****************************************************************************************
static unsigned char HttpSendResponseHeader(const unsigned char connectionId, const HttpStatus *status, unsigned char *headers, unsigned short headersLength, unsigned short dataLength){
 if(incomingRequestState == HTTP_REQUEST_STATE_NO_REQUEST){
  return 0;
 }
 incomingRequestState = HTTP_REQUEST_STATE_START_REQUEST;
 unsigned char staticHeaders[52];
 int printResult = snprintf(staticHeaders, 50, "HTTP/1.0 %u %s", status->code, status->message ?: "Shit Happens");
 if(printResult < 0 || printResult >= 50){
  TcpDisconnect(connectionId, 5000);
  return 0;
 }
 strcat(staticHeaders, HTTP_HEADER_ROW_BREAK);
 if(!TcpSendData(connectionId, 60000, staticHeaders, strlen(staticHeaders))){
  TcpDisconnect(connectionId, 5000);
  return 0;
 }
 if(headersLength){
  if(!TcpSendData(connectionId, 60000, headers, headersLength)){
   TcpDisconnect(connectionId, 5000);
   return 0;
  }
 }
 snprintf(staticHeaders, 52, "Connection: close" HTTP_HEADER_ROW_BREAK "Content-Length: %u" HTTP_HEADER_ROW_BREAK HTTP_HEADER_ROW_BREAK, dataLength);
 if(!TcpSendData(connectionId, 60000, staticHeaders, strlen(staticHeaders))){
  TcpDisconnect(connectionId, 5000);
  return 0;
 }
 if(dataLength == 0){
  TcpDisconnect(connectionId, 5000);
 }
 return 1;
}

//*****************************************************************************************
//
// Function : HttpParseRequestHeader
// Description : function take char from incoming data and decide where put them in request structure on fail send error response header
// return 1 on success 0 on any error
//
//*****************************************************************************************
static unsigned char HttpParseRequestHeader(const unsigned char ch){
 if(incomingRequestState == HTTP_REQUEST_STATE_START_REQUEST){
  incomingRequestState = HTTP_REQUEST_STATE_METHOD;
  incomingPosition = 0;
 }
 // parse http method
 if(incomingRequestState == HTTP_REQUEST_STATE_METHOD){
  if(ch == ' '){
   incomingRequestState = HTTP_REQUEST_STATE_URL;
   incomingRequest.method[incomingPosition] = 0;
   incomingRequest.urlLength = 0;
   return 1;
  }
  if(ch < 'A' || ch > 'Z'){
   HttpStatus status = {400, "Header Syntax Error"};
   HttpSendResponseHeader(incomingRequestConnectionId, &status, 0, 0, 0);
   return 0;
  }
  if(incomingPosition >= HTTP_MAX_METHOD_LENGTH){
   HttpStatus status = {431, "Header Part Too Large"};
   HttpSendResponseHeader(incomingRequestConnectionId, &status, 0, 0, 0);
   return 0;
  }
  incomingRequest.method[incomingPosition] = ch;
  incomingPosition++;
  return 1;
 }
 // parse url
 if(incomingRequestState == HTTP_REQUEST_STATE_URL){
  if(ch == ' '){
   incomingRequestState = HTTP_REQUEST_STATE_VERSION;
   incomingPosition = 0;
   return 1;
  }
  if(incomingRequest.urlLength >= HTTP_MAX_URL_LENGTH){
   HttpStatus status = {414, "URI Too Long"};
   HttpSendResponseHeader(incomingRequestConnectionId, &status, 0, 0, 0);
   return 0;
  }
  incomingRequest.url[incomingRequest.urlLength] = ch;
  incomingRequest.urlLength++;
  return 1;
 }
 // parse http version
 if(incomingRequestState == HTTP_REQUEST_STATE_VERSION){
  if(ch == '\r' || ch == '\n'){
   incomingMessage.version[incomingPosition] = 0;
   incomingMessage.headersLenght = 0;
   if(ch == '\r'){
    incomingRequestState = HTTP_STATE_MAC_END_HEADER;
   }
   if(ch == '\n'){
    incomingRequestState = HTTP_STATE_LINUX_END_HEADER;
   }
   return 1;
  }
  if(incomingPosition >= HTTP_MAX_VERSION_LENGTH){
   HttpStatus status = {431, "Header Part Too Large"};
   HttpSendResponseHeader(incomingRequestConnectionId, &status, 0, 0, 0);
   return 0;
  }
  incomingMessage.version[incomingPosition] = ch;
  incomingPosition++;
  return 1;
 }
 // parse rest header rows
 if(incomingRequestState >= HTTP_STATE_LINUX_END_HEADER && incomingRequestState <= HTTP_STATE_HEADER){
  if(!HttpParseHeader(ch, &incomingRequestState)){
   HttpStatus status = {431, "Header Part Too Large"};
   HttpSendResponseHeader(incomingRequestConnectionId, &status, 0, 0, 0);
   return 0;
  }
  return 1;
 }
 HttpStatus status = {400, "Header Syntax Error"};
 HttpSendResponseHeader(incomingRequestConnectionId, &status, 0, 0, 0);
 return 0;
}

//*****************************************************************************************
//
// Function : HttpInit
// Description : function set memory into init state
//
//*****************************************************************************************
void HttpInit(){
 incomingRequestState = HTTP_REQUEST_STATE_NO_REQUEST;
 incomingRequestConnectionId = TCP_INVALID_CONNECTION_ID;
 incomingRequest.message = &incomingMessage;
}

//*****************************************************************************************
//
// Function : HttpParseHeaderValue
// Description : function parse header value from message
// return HttpHeaderValue structure with 0 pointer value property if header not found or header row have wrong syntax
//
//*****************************************************************************************
const HttpHeaderValue HttpParseHeaderValue(const HttpMessage *message, const unsigned char *header){
 HttpHeaderValue headerValue;
 headerValue.value = strstr(message->headers, header);
 if(!headerValue.value){
  headerValue.length = 0;
  return headerValue;
 }
 headerValue.value += strlen(header);
 headerValue.length = message->headersLenght - (headerValue.value - message->headers);
 if(!headerValue.length){
  headerValue.value = 0;
  return headerValue;
 }
 if(headerValue.value[0] != ':'){
  headerValue.value = 0;
  headerValue.length = 0;
  return headerValue;
 }
 if(headerValue.length >= 2 && headerValue.value[1] == ' '){
  headerValue.value += 2;
  headerValue.length -= 2;
 }else{
  headerValue.value++;
  headerValue.length--;
 }
 unsigned short i;
 for(i=0; i<headerValue.length; i++){
  if(headerValue.value[i] == '\n' || headerValue.value[i] == '\r'){
   headerValue.length = i;
   break;
  }
 }
 return headerValue;
}

//*****************************************************************************************
//
// Function : HttpSendResponse
// Description : function send response for client, can be used only once per request handling
//
//*****************************************************************************************
unsigned char HttpSendResponse(const HttpStatus *status, unsigned char *headers, unsigned short headersLength, unsigned char *data, unsigned short dataLength){
 if(incomingRequestState != HTTP_REQUEST_STATE_REQUEST_HANDLING){
  return 0;
 }
 unsigned char connectionId = incomingRequestConnectionId;
 if(!HttpSendResponseHeader(connectionId, status, headers, headersLength, dataLength)){
  return 0;
 }
 if(dataLength == 0){
  return 1;
 }
 if(!TcpSendData(connectionId, 60000, data, dataLength)){
  TcpDisconnect(connectionId, 5000);
  return 0;
 }
 TcpDisconnect(connectionId, 5000);
 return 1;
}

//*****************************************************************************************
//
// Function : HttpTcpOnNewConnection
// Description : defined tcp callback for accept new incoming connection on http port
//
//*****************************************************************************************
unsigned char HttpTcpOnNewConnection(const unsigned char connectionId){
 const TcpConnection *connection = TcpGetConnection(connectionId);
 if(connection->port != HTTP_SERVER_PORT){
  #if HTTP_TCP_INCLUDED == 1
  return TcpOnNewConnection(connectionId);
  #else
  return NET_HANDLE_RESULT_DROP;
  #endif
 }
 if(incomingRequestState != HTTP_REQUEST_STATE_NO_REQUEST){
  return NET_HANDLE_RESULT_DROP;
 }
 incomingRequestConnectionId = connectionId;
 incomingRequestState = HTTP_REQUEST_STATE_START_REQUEST;
 incomingPosition = 0;
 incomingRequest.connection = connection;
 return NET_HANDLE_RESULT_OK;
}

//*****************************************************************************************
//
// Function : HttpTcpOnConnect
// Description : defined tcp callback for detect new connection not used in http
//
//*****************************************************************************************
void HttpTcpOnConnect(const unsigned char connectionId){
 #if HTTP_TCP_INCLUDED == 1
 if(connectionId == incomingRequestConnectionId){
  return;
 }
 TcpOnConnect(connectionId);
 #endif
}

//*****************************************************************************************
//
// Function : HttpTcpOnIncomingData
// Description : handle incoming http request by tcp callback
//
//*****************************************************************************************
void HttpTcpOnIncomingData(const unsigned char connectionId, const unsigned char *data, unsigned short dataLength){
 if(connectionId != incomingRequestConnectionId){
  #if HTTP_TCP_INCLUDED == 1
  TcpOnIncomingData(connectionId, data, dataLength);
  #endif
  return;
 }
 if(incomingRequestState == HTTP_REQUEST_STATE_NO_REQUEST || connectionId != incomingRequestConnectionId){
  TcpDisconnect(connectionId, 5000);
  return;
 }
 // parse request headers
 unsigned short dataPosition = 0;
 if(incomingRequestState < HTTP_STATE_END_HEADER){
  for(dataPosition=0; dataPosition<dataLength; dataPosition++){
   if(!HttpParseRequestHeader(data[dataPosition])){
    return;
   }
   if(incomingRequestState == HTTP_STATE_END_HEADER){
    dataPosition++;
    break;
   }
  }
  if(incomingRequestState != HTTP_STATE_END_HEADER){
   return;
  }
 }
 // read request body
 if(incomingRequestState < HTTP_STATE_END_MESSAGE){
  if(strcmp(incomingRequest.method, "HEAD") == 0 || strcmp(incomingRequest.method, "GET") == 0 || strcmp(incomingRequest.method, "OPTIONS") == 0){
   incomingRequestState = HTTP_STATE_END_MESSAGE;
  }else{
   const HttpHeaderValue contentLength = HttpParseHeaderValue(&incomingMessage, "Content-Length");
   if(!contentLength.value){
    HttpStatus status = {411, "Length Required"};
    HttpSendResponseHeader(connectionId, &status, 0, 0, 0);
    return;
   }
   unsigned long httpDataLength;
   if(!ParseLong(&httpDataLength, contentLength.value, contentLength.length)){
    HttpStatus status = {400, "Header Syntax Error"};
    HttpSendResponseHeader(connectionId, &status, 0, 0, 0);
    return;
   }
   if(httpDataLength > HTTP_MAX_DATA_LENGTH){
    HttpStatus status = {413, "Payload Too Large"};
    HttpSendResponseHeader(connectionId, &status, 0, 0, 0);
    return;
   }
   HttpMessagePutData(data, dataLength, dataPosition);
   if(incomingMessage.dataLength == httpDataLength){
    incomingRequestState = HTTP_STATE_END_MESSAGE;
   }else{
    return;
   }
  }
 }
 // handle request and send response
 if(incomingRequestState == HTTP_STATE_END_MESSAGE){
  incomingRequestState = HTTP_REQUEST_STATE_REQUEST_HANDLING;
  HttpOnIncomingRequest(&incomingRequest);
  if(incomingRequestState == HTTP_REQUEST_STATE_REQUEST_HANDLING){
   TcpDisconnect(connectionId, 5000);
   return;
  }
  return;
 }
 HttpStatus status = {400, "Request Is In Invalid State"};
 HttpSendResponseHeader(connectionId, &status, 0, 0, 0);
 return;
}

//*****************************************************************************************
//
// Function : HttpTcpOnDisconnect
// Description : after handling request reset memory and be ready for another http request
//
//*****************************************************************************************
void HttpTcpOnDisconnect(const unsigned char connectionId){
 if(connectionId != incomingRequestConnectionId){
  #if HTTP_TCP_INCLUDED == 1
  TcpOnDisconnect(connectionId);
  #endif
  return;
 }
 if(incomingRequestState == HTTP_REQUEST_STATE_NO_REQUEST){
  return;
 }
 incomingRequestConnectionId = TCP_INVALID_CONNECTION_ID;
 incomingRequestState = HTTP_REQUEST_STATE_NO_REQUEST;
}
#endif
